kube-runtime/src/reflector/shared_store.rs

```diff
@@ -0,0 +1,401 @@
+use crate::{
+    reflector::{
+        reflector,
+        store::{Store, Writer},
+    },
+    utils::{stream_subscribe, StreamSubscribe, WatchStreamExt},
+    watcher::{self, watcher, Event},
+};
+use futures::{stream, Stream, StreamExt};
+use k8s_openapi::NamespaceResourceScope;
+use kube_client::{api::ListParams, Api, Client, Resource};
+use serde::de::DeserializeOwned;
+use std::{collections::HashMap, fmt::Debug, hash::Hash, sync::Arc};
+use stream::BoxStream;
+
+pub struct SharedStore<K, W = WatcherFactory<K>>
+where
+    K: 'static + Resource + Clone + DeserializeOwned + Debug + Send,
+    K::DynamicType: Hash + Eq,
+    W: CreateWatcher<K>,
+{
+    watcher_provider: W,
```

Comment:
```suggestion
pub struct SharedStore<K>
where
    K: 'static + Resource + Clone + DeserializeOwned + Debug + Send,
    K::DynamicType: Hash + Eq,
{
```
I feel like we can get away with not having a second generic parameter for this. The `WatcherFactory`'s impls have exactly the same constraints as the `SharedStore` methods that construct them. Did you have a particular use case in mind for the `WatcherFactory`?

Reply:
It's purely for the tests. I don't particularly like test induced damage to non-test code like this ... but I also don't like not unit testing stuff 😅. Do you have a better suggestion on how to test something like this?

Reply:
Possibly worth noting, users cannot create an instance of the `SharedStore` where `W` is anything other than a `WatcherFactory` since non of the properties are public and the only constructor we expose is for the variation using a `WatcherFactory`. I don't _think_ end users ever need to know or care about the `W` generic 🤔.

Reply:
hm, yeah, it's probably not that bad, I would also ideally have the damage confined to the tests if possible. was thinking there might be some ways to mock just one of the functions (like setup_reflector) used in this module, but maybe it's only [mocktopus](https://github.com/CodeSandwich/Mocktopus) that lets you do functions (and that's nightly only).

i think it probably is OK. mocking watch streams is hard to do raw. i would otherwise suggest the mocked client with tower_test as is used in client/api.